kernel/src/hil/digest.rs

```diff
@@ -0,0 +1,62 @@
+//! Interface for Digest
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// This callback is called when the data has been added to the digest
+    /// engine.
+    /// On error `result` will contain a return code
+    /// On error or success `digest` will contain a reference to the original
+    /// data supplied to `add_data()`.
+    fn add_data_done(&'a self, result: Result<(), ReturnCode>, data: &'static mut [u8]);
+
+    /// This callback is called when a digest is computed.
+    /// On error `result` will contain device specific error codes
+    /// On error or success `digest` will contain a reference to the original
+    /// digest supplied to `run()`.
+    fn hash_done(&'a self, result: Result<(), u32>, digest: &'static mut [u8]);
+}
+
+pub trait DigestType: Eq + Copy + Clone + Sized + AsRef<[u8]> + AsMut<[u8]> {}
+
+impl DigestType for [u8; 32] {}
+
+/// Computes a digest (cryptographic hash) over data
+pub trait Digest<'a, T: DigestType> {
+    /// Set the client instance which will receive `hash_done()` and
+    /// `add_data_done()` callbacks.
+    /// This callback is called when the data has been added to the digest
+    /// engine.
+    /// The callback should follow the `Client` `add_data_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Add data to the digest block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes parsed on success
+    /// There is no guarentee the data has been written until the `add_data_done()`
+    /// callback is fired.
+    /// On error the return value will contain a return code and the original data digest
+    fn add_data(&self, data: &'static mut [u8]) -> Result<usize, (ReturnCode, &'static mut [u8])>;
```

Comment:
I think this needs to take a length as well (or use a LeaseableBuffer https://github.com/tock/tock/blob/master/kernel/src/common/leasable_buffer.rs). The issue is if a user wants to add only a portion of a buffer to a digest, it either has to have a buffer exactly that length, or provide only a slice to Digest. If it only provides a slice, then it will only ever get back that slice, and looses access to the entire buffer.

Reply:
I don't see why the `data` slice should be mutable here.
And if it isn't mutable, the caller can pass a slice without worrying about passing an extra length nor loosing the slice.

Reply:
But then the caller wouldn't be able to hold on to the mutable reference as well. `data` is `&mut` so that the HMAC implementor can pass back a mutable reference when it is done with the buffer.

Reply:
I have updated this to use `LeasableBuffer`